dsc_lib/src/progress.rs

```diff
@@ -0,0 +1,192 @@
+    pub id: String,
+    /// The activity being performed.
+    pub activity: Option<String>,
+    /// The percentage of the operation that is complete from 0 to 100.
+    #[serde(rename = "percentComplete")]
+    pub percent_complete: u8,
+    /// The status of the operation.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub status: Option<String>,
+    /// The number of seconds remaining in the operation.
+    #[serde(rename = "secondsRemaining", skip_serializing_if = "Option::is_none")]
+    pub seconds_remaining: Option<u64>,
+    /// The name of the resource being operated on.
+    #[serde(rename = "resourceName", skip_serializing_if = "Option::is_none")]
+    pub resource_name: Option<String>,
+    /// The type of the resource being operated on.
+    #[serde(rename = "resourceType", skip_serializing_if = "Option::is_none")]
+    pub resource_type: Option<String>,
+    /// The result of the operation.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub result: Option<Value>,
+}
+
+impl Progress {
+    #[must_use]
+    pub fn new() -> Progress {
+        Progress {
+            id: Uuid::new_v4().to_string(),
+            ..Default::default()
+        }
+    }
+}
+
+pub struct ProgressBar {
+    progress_value:  Progress,
+    console_bar: Span,
+    item_count: u64,
+    item_position: u64,
+    format: ProgressFormat
+}
+
+impl ProgressBar {
+    /// Create a `ProgressBar` object to update progress
+    ///
+    /// # Arguments
+    ///
+    /// * `item_count` - Total number of steps to complete.  Use '1' if unknown and increment when complete.
+    /// * `format` - The `ProgressFormat` to render.
+    ///
+    /// # Returns
+    ///
+    /// A `ProgressBar` oject to update progress
+    ///
+    /// # Errors
+    ///
+    /// Fails if progress style for console rendering can't be set.
+    ///
+    pub fn new(item_count: u64, format: ProgressFormat) -> Result<ProgressBar, DscError> {
+        let bar = warn_span!("");
+        if format == ProgressFormat::Default {
+            bar.pb_set_style(&ProgressStyle::with_template(
+                "{spinner:.green} [{elapsed_precise:.cyan}] [{bar:40.cyan/blue}] {pos:>7}/{len:7} {msg:.yellow}"
+            )?);
+            bar.pb_set_length(item_count);
+            let _guard = bar.enter();
+        }
+
+        Ok(ProgressBar {
+            progress_value: Progress::new(),
+            console_bar: bar,
+            item_count,
+            item_position: 0,
+            format
+        })
+    }
+
+    pub fn increment(&mut self, delta: u64) {
+        if self.format == ProgressFormat::None {
+            return;
+        }
+
+        self.item_position += delta;
+
+        if self.item_count  > 0 {
+            self.progress_value.percent_complete = if self.item_position >= self.item_count {
+                100
+            } else {
+                u8::try_from((self.item_position * 100) / self.item_count).unwrap_or(100)
+            };
+        }
```

Comment:
Can this be a helper function since its repeated on L. 150 and L. 169? Maybe with an option to `write_json()`?

Reply:
Good idea